include/swift/SILOptimizer/Analysis/EscapeAnalysis.h

```diff
@@ -256,28 +256,29 @@ class EscapeAnalysis : public BottomUpIPAnalysis {
   };
 
 public:
-  
   /// Indicates to what a value escapes. Note: the order of values is important.
   enum class EscapeState : char {
 
     /// The node's value does not escape.
     /// The value points to a locally allocated object who's lifetime ends in
     /// the same function.
     None,
-    
+
     /// The node's value escapes through the return value.
     /// The value points to a locally allocated object which escapes via the
     /// return instruction.
     Return,
 
-    /// The node's value escapes through a function argument.
-    Arguments,
+    /// The node's value escapes through a non-guaranteed function argument.
+    AnyArguments,
+
+    /// The node's value escapes through a guaranteed function argument.
+    GuaranteedArguments,
```

Comment:
What if it escapes via both. Which state is more conservative?

Reply:
That's a good point. `AnyArguments` should take priority.

Reply:
I think it should go `None` then `GuaranteedArguments` then `Return` then `AnyArguments`. WDYT?

Reply:
Most of the queries don't care if pointers are returned, since they only care about aliasing within the current function. But if a pointer escapes via an argument then it escapes in probably every way that matters regardless of the convention. The EscapeState refers to the content pointed-to by the pointer or reference, not the pointer itself. So if an object escapes via a guaranteed reference, that object can still be deleted by some unknown code within the same function. It's not clear yet how you could make use of this distinction..

Reply:
Are you saying that a class type could still be destroyed in another function even if it's passed to that function as a guaranteed parameter? 

Here's how I'd like to make use of this distinction (which I can't do if the above statement is true): in stack promotion, if the only use of an object is as a guaranteed parameter, then we can still promote it to a stack alloc because we know all the points at which it will be destroyed.

Reply:
I'd need to look at the stack promotion example, but, yes, if the object escapes (even via a guaranteed argument) there's no way to know the object's lifetime.

Reply:
Isn't the point of a guaranteed argument that the object is live when it enters and live when it exits? Is this different for reference types?

Reply:
An object does not need to be passed as an argument to a function to be destroyed by that function

Reply:
Hmm, I see. [Here's an edge case](https://swift.godbolt.org/z/R9_vXX) that doesn't work. The guaranteed argument is assigned to a global and the global is released, because it's a pointer, it also releases the argument. 

I'll keep thinking about ways we could do this optimization, I think it would be very beneficial.